### ISSUE TITLE
crontab -l should enumerate all cronjobs

### ISSUE THREAD

**Issue by ceztko** [cronie-crond/cronie#76]
`crontab -l [-u]` should enumerate all the jobs for the user, including all entries in various `/etc/cron.*/*` .  Only the daemon itself can accurately report its internal state, without possible race conditions, so one should not rely on manually parsing entries in /etc. The issue has been reported also on RedHat [bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1759231) (not much is moving there, but at least the problem seems to be acknowledged) and seems to have been always present since the advent of `/etc/cron.*/*`. Also there should be a mean to immediately refresh the configuration, without waiting for the file changes notification.

**Comment by khardix** [comment]
This might take a while. As the assignee for the [RedHat bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1759231), I took a look at the current state:

1.  The `crontab -l` invocation does not consult the running daemon at all; currently, it just dumps `/var/spool/cron/$USER` to stdout (if the file exists). Parsing of other files could probably be added without too much effort, but that is essentially the built-in equivalent of "manually parsing entries in /etc".

    Also note that the other `crontab` manipulation options (from what I've checked, at least `-d`) suffer from the same issue – they only manipulate the file on disk.

2.  The issue description lead me to believe that @ceztko considers the `crontab` command to be a specialized client to the running `crond` process, which unfortunately seems to be a false assumption. We could probably try to modify it to be one; however, it would be a bigger design endeavour. Is anybody aware of any prior work in this are (i.e. in other cron implementations)? Also, is this something you would even accept, @t8m?

**Comment by t8m** [comment]
I am not fond of this RFE much. You described the problems with it fairly precisely.
One thing is clear, we should not change the behavior of existing crontab -l, it would have to be an additional option (to be used along the -l).
There is a cronnext that perhaps could be an inspiration for an implementation, or perhaps cronnext could be enhanced to support the use case that triggered the report of this RFE.

**Comment by ceztko** [comment]
> 1. The `crontab -l` invocation does not consult the running daemon at all; currently, it just dumps `/var/spool/cron/$USER` to stdout (if the file exists). Parsing of other files could probably be added without too much effort, but that is essentially the built-in equivalent of "manually parsing entries in /etc".

Wow, the Unix philosophy of "doing one thing and doing it well" has been taken with a little bit of freedom here.

> 2. The issue description lead me to believe that @ceztko considers the `crontab` command to be a specialized client to the running `crond` process, which unfortunately seems to be a false assumption.

The lack of any IPC between cron and some specialized clients is sad news. How one can move forward and implement some more useful features without such mechanism? For example, let's assume it exists something like a `cronctl` command that implement it. Some may want to do something like `cronctl refresh` to immediately perform a reading configuration files and print syntax errors in the output (not in the log) and without waiting for the next cycle. Or, assuming a way to label cron jobs is added, one may want to run a job immediately like `cronctl run jobname`, also printing errors on the output that are hard to spot because of environment differences between a regular console session and cron daemon executing scripts.

> I am not fond of this RFE much.

But exactly why you are not? The use case for such features is debugging: in system administration tasks it happened to me multiple times that I was 100% convinced a task was scheduled when in fact it was not. Seeing the precise state of the running daemon would have saved me lot of headaches. Just ignore the GUI, but by comparison with Windows Scheduler the latter seems to expose such kind of interface to communicate with the internal state of the daemon. Of course it's a lot of work to do to add such features.

![image](https://user-images.githubusercontent.com/3037449/116895285-512ac880-ac33-11eb-9adc-81982f96a73d.png)

**Comment by t8m** [comment]
Exactly because the reason you've described. To properly implement it requires IPC between crond and the fictional cronctl command which is definitely not a small feature. I personally do not have time to implement it. If someone wants to take on this one feel free, but I'd suggest discussing a design and implemention proposal in a separate issue first.

> The lack of any IPC between cron and some specialized clients is sad news.

This is definitely no news at all. :grin: It was always like this to keep the crond simple.

But my first proposal for anybody trying to implement something like this would be to look at cronnext whether it would not at least partially suffice. Of course with the disadvantage that it would not reflect the current crond state but the state as if the crond was restarted at this point.

**Comment by khardix** [comment]
I will have to side with @t8m – addition of IPC mechanism and/or `cronctl` looks like a pretty big feature creep. Given on how ubiquitous the daemon is, I would be very wary of modifying an existing project to grow such features.

As a side note, if you really want the feature described, I would suggest to look at systemd timers – they should already have the ability to list all scheduled jobs and run them on demand.

**Comment by ceztko** [comment]
@khardix I actually never used systemd timers, although I actually searched for such features in systemd some time ago. They seems to be a good match for what I need. At this point, if daemons like cronie are kept just for compatibility, a decent fix for this exact issue (without caring about advanced use cases that can be solved only with IPC) would be just to do something similar of what `crontab -l` is doing right now outputting content of `/etc/cron.*/*`, possibly with a little bit of validation.

**Comment by khardix** [comment]
> a decent fix for this exact issue (without caring about advanced use cases that can be solved only with IPC) would be just to do something similar of what crontab -l is doing right now outputting content of /etc/cron.*/*, possibly with a little bit of validation

That is something that I think would be in scope. I will take a closer look and try to prepare a patch for that, using cronnext implementation as a reference.

**Comment by t8m** [comment]
I wonder if cronnext by itself already does not fulfill the use case. In particular `cronnext -c` command might be of an interest.

**Comment by khardix** [comment]
So, I was toying with this on and off for a few weeks now, and the more I think about this request, the less sense it makes (at least to me).

TL;DR: For listing of all current entries in consistent format, `cronnext -c` is probably the way to go; for anything else, you would probably be better off actually reading the files from disk in a shell script.

Longer version: `crontab -l` is supposed to list crontab for a specific user; what's more, it only ever deals with `/var/spool/cron/*` files – there is not really a way to instruct it to interact with the system tabs (`/etc/crontab`, `/etc/cron.d/*`). In this regard, the current implementation is consistent with the rest of the options – it just dumps the `/var/spool/cron/{user}` file to stdout.

If I would try to expand it to also read (and possibly filter by user) the system crontabs, I would have to figure out how to diferentiate/delimit the files the jobs came from, because they do not inherit execution environment from one another; coding the equivalent of `grep $user /etc/crontab /etc/cron.d/*` is not really gonna work so well.

Alternatively, I could dump all the jobs in consistent, machine-readable format with all relevant information (environment, etc.) for each job attached; it wouldn't look like a crontab, but would have all the available info. Turns out, that's what `cronnext -c` does (although it seems to not list the environment; that could possibly be added):

```
$ cronnext -c
spool: /var/spool/cron
crontabs:
  - user: "*system*"
    crontab: /etc/crontab
    system: 1
    entries:
  - user: "*system*"
    crontab: /etc/cron.d/0hourly
    system: 1
    entries:
      - user: root
        cmd: "run-parts /etc/cron.hourly"
        flags: 0x0E
        flagnames: HR_STAR|DOM_STAR|DOW_STAR
        delay: 0
        next: 1626966060
        nextstring: Thu Jul 22 17:01:00 2021
next: 1626966060
```

So, I'm inclined to consider this a good learning experience in cronie internals and codebase and to close this (and possibly the [original bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1759231) as well) without changing anything.

@ceztko, would `cronnext -c` work for you?

**Comment by khardix** [comment]
No response so far; in that case, I would consider this feature is no longer necessary or needed. I'm going to close the original bugzilla; @t8m, feel free to close this as well.

**Comment by t8m** [comment]
Closing.

**Comment by ceztko** [comment]
Considering the existence of `cronnext -c`, I'm ok with with the closing the ticket. The unfortunate situation is that `cronnext` is not ubiquitous: if you search for "list user cron jobs" or similar chances are that first results will just point to `crontab -l`, so having the feature directly in `crontab` would have offered a better user experience and be an example to imitate in other cron implementations.
